### PR TITLE
Escape quotes in headers correctly in all languages

### PR DESCRIPTION
**Pull request by pimterry**
As discussed in https://github.com/Kong/httpsnippet/issues/271#issuecomment-1185716335 - this ports the test changes from https://github.com/httptoolkit/httpsnippet into the new folder structure here, providing working examples for correct quote-escaping for every single language/format.

~~This does _not_ make the tests pass, we'll need to do that separately - currently this fails for 22 of the snippets, where they're currently not escaping quotes right now.~~

~~I think the base logic to fix this should look something like [this](https://github.com/httptoolkit/httpsnippet/blob/a10703297c8b03cbf62b8c5d282cbdf36596fa04/src/helpers/format.js#L3-L45), which we'll need to call in the appropriate places with the appropriate options (delimiter, escapeChar and escapeNewlines). You can see how it's used for each snippet [in the original commit](https://github.com/httptoolkit/httpsnippet/commit/a10703297c8b03cbf62b8c5d282cbdf36596fa04) to check which ones you need, but the tests here should be enough to confirm that's correct.~~

~~Happy for somebody else to look into fixing this, if anybody has time, or I should be able to migrate the fix for this myself late this week or early next if not.~~

EDIT: Now updated to add a working implementation, in addition to updating all the test fixtures to cover this.

### CODE REVIEW

**Comment by dimitropoulos** [comment]
wow thanks!  sounds like you can see we're catching up around here! :)

Still working on the readme, but there's a super useful shortcut I wanted to tell you about: `OVERWRITE_EVERYTHING=true npm run test`.  Like it sounds, haha, it will overwrite all tests.  It's useful when working on a change that will update fixtures.

**Comment by pimterry** [comment]
All done, this now fixes #271, and there's also a tiny patch in here for weird indentation around python3 header values.

File: src/helpers/escape.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * for the complete original algorithm.
+ */
+export function escapeString(value: any, options: EscapeOptions = {}) {
+  const {
+    delimiter = '"',
+    escapeChar = '\\',
+    escapeNewlines = true
+  } = options;
+
+  value = value.toString();
```
Comment (dimitropoulos) [inline src/helpers/escape.ts]: nit: if it's easy to avoid (and it seems like it is in this case) we should avoid parameter reassignment (e.g. https://eslint.org/docs/latest/rules/no-param-reassign).  thoughts?

File: src/helpers/escape.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    } else {
+      return c;
+    }
+  }).join('');
+}
+
+export const escapeForSingleQuotes = (value: any) =>
+  escapeString(value, { delimiter: "'" });
+
+export const escapeForDoubleQuotes = (value: any) =>
```
Comment (dimitropoulos) [inline src/helpers/escape.ts:94]: could you please add a TSDoc here explaining in layman's terms what this function does?

File: src/helpers/escape.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      // using this primarily for header values that are generally (though not 100%
+      // strictly?) ASCII-only, this should almost never happen.
+      return JSON.stringify(c).slice(1, -1);
+    } else {
+      return c;
+    }
+  }).join('');
+}
+
+export const escapeForSingleQuotes = (value: any) =>
```
Comment (dimitropoulos) [inline src/helpers/escape.ts:84]: could you please add a TSDoc here explaining in layman's terms what this function does?

File: src/targets/python/python3/client.ts [hunk 4]
```diff
@@ -65,13 +66,13 @@ export const python3: Client<Python3Options> = {
...
         if (count++ !== headerCount) {
-          push(`    '${header}': "${headers[header]}",`);
+          push(`    '${header}': "${escapeForDoubleQuotes(headers[header])}",`);
         } else {
-          push(`    '${header}': "${headers[header]}"`);
+          push(`    '${header}': "${escapeForDoubleQuotes(headers[header])}"`);
         }
       }
 
-      push('    }');
```
Comment (dimitropoulos) [inline src/targets/python/python3/client.ts:74]: good catch

File: src/helpers/escape.test.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import { escapeString } from './escape';
+
+describe('Escape methods', () => {
```
Comment (dimitropoulos) [inline src/helpers/escape.test.ts:3]: thank you for the unit tests in addition to all the fixtures!

**Review by dimitropoulos — APPROVED** [review]
@erunion could you please take a look?  otherwise, I'm happy with this!

@pimterry you're the first brave soul to update any of the root fixtures!  thanks!  If there is anything about the process that was needlessly difficult for you, please let me know. :)

**Review by erunion — APPROVED** [review]

File: src/helpers/escape.ts (hunk 1 above)
Comment (pimterry) [reply src/helpers/escape.ts]: Sure, seems reasonable! Done.

File: src/helpers/escape.ts (hunk 2 above)
Comment (pimterry) [reply src/helpers/escape.ts:94]: Done :+1:

File: src/helpers/escape.ts (hunk 3 above)
Comment (pimterry) [reply src/helpers/escape.ts:84]: Done :+1:
